component/org.wso2.carbon.identity.dpop/src/main/java/org/wso2/carbon/identity/dpop/handler/DPoPAuthenticationHandler.java

```diff
@@ -0,0 +1,106 @@
+import org.wso2.carbon.identity.dpop.validators.DPoPValidator;
+import org.wso2.carbon.identity.oauth2.OAuth2TokenValidationService;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2ClientApplicationDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationRequestDTO;
+import org.wso2.carbon.identity.oauth2.dto.OAuth2TokenValidationResponseDTO;
+
+/**
+ * DPoPAuthenticationHandler will validate the requests authorized with DPoP access tokens.
+ */
+public class DPoPAuthenticationHandler extends AuthenticationHandler {
+
+    private static final Log log = LogFactory.getLog(DPoPAuthenticationHandler.class);
+
+    @Override
+    protected AuthenticationResult doAuthenticate(MessageContext messageContext) throws
+            AuthenticationFailException {
+
+        AuthenticationResult authenticationResult = new AuthenticationResult(AuthenticationStatus.FAILED);
+        AuthenticationContext authenticationContext = (AuthenticationContext) messageContext;
+        AuthenticationRequest authenticationRequest = authenticationContext.getAuthenticationRequest();
+        if (authenticationRequest != null) {
+
+            String authorizationHeader = authenticationRequest.getHeader(HttpHeaders.AUTHORIZATION);
+            if (StringUtils.isNotBlank(authorizationHeader) &&
+                    authorizationHeader.startsWith(DPoPConstants.OAUTH_DPOP_HEADER) ) {
+                String accessToken;
+                String[] bearerToken = authorizationHeader.split(" ");
+                if (bearerToken.length == 2) {
+                    accessToken = bearerToken[1];
+                    OAuth2TokenValidationService oAuth2TokenValidationService = new OAuth2TokenValidationService();
+                    OAuth2TokenValidationRequestDTO requestDTO = new OAuth2TokenValidationRequestDTO();
+                    OAuth2TokenValidationRequestDTO.OAuth2AccessToken token = requestDTO.new OAuth2AccessToken();
+                    token.setIdentifier(accessToken);
+                    token.setTokenType(DPoPConstants.OAUTH_HEADER);
+                    requestDTO.setAccessToken(token);
+
+                    OAuth2TokenValidationRequestDTO.TokenValidationContextParam contextParam = requestDTO.new
+                            TokenValidationContextParam();
+
+                    OAuth2TokenValidationRequestDTO.TokenValidationContextParam[] contextParams = { contextParam };
+                    requestDTO.setContext(contextParams);
+
+                    OAuth2ClientApplicationDTO clientApplicationDTO =
+                            oAuth2TokenValidationService.findOAuthConsumerIfTokenIsValid(requestDTO);
+                    OAuth2TokenValidationResponseDTO responseDTO =
+                            clientApplicationDTO.getAccessTokenValidationResponse();
+
+                    DPoPValidator.validateDPoPHeaderandToken(authenticationResult,responseDTO,
+                            authenticationRequest);
+
+                } else {
```

Comment:
Remove these nested if else part and move this else part to top for better readability

Reply:
Fixed with [23dad1a](https://github.com/kalanika/identity-oauth-addons/commit/23dad1add37a341343e363c8920846dbbe961d0a)